File: Y2021/Day9.java
package AdventOfCode.Y2021;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;

public class Day9 {

    static int[][] caves = new int[100][100];
    static boolean[][] visited;

    public static void part1() {
        int sum = 0;

        for (int i=0; i<caves.length; i++) {
            for (int j=0; j<caves[0].length; j++) {
                boolean isLowest = true;
                if (i!=0 && caves[i][j] >= caves[i-1][j]) isLowest = false;
                if (i!=caves.length-1 && caves[i][j] >= caves[i+1][j]) isLowest = false;
                if (j!=0 && caves[i][j] >= caves[i][j-1]) isLowest = false;
                if (j!=caves[0].length-1 && caves[i][j] >= caves[i][j+1]) isLowest = false;

                if (isLowest) sum += caves[i][j] + 1;
            }
        }

        System.out.println(sum);
    }

    public static void part2() {
        ArrayList<Integer[]> coordsList = new ArrayList<>(); // find all low points
        for (int i=0; i<caves.length; i++) {
            for (int j=0; j<caves[0].length; j++) {
                boolean isLowest = true;
                if (i!=0 && caves[i][j] >= caves[i-1][j]) isLowest = false;
                if (i!=caves.length-1 && caves[i][j] >= caves[i+1][j]) isLowest = false;
                if (j!=0 && caves[i][j] >= caves[i][j-1]) isLowest = false;
                if (j!=caves[0].length-1 && caves[i][j] >= caves[i][j+1]) isLowest = false;

                if (isLowest) coordsList.add(new Integer[]{i,j});
            }
        }

        int numcoords = coordsList.size();
        int[] countsPerBasin = new int[numcoords];

        for (int i=0; i<numcoords; i++) {
            visited = new boolean[100][100];
            countsPerBasin[i] = basinFill(coordsList.get(i)[0], coordsList.get(i)[1]);
        }
        
        Arrays.sort(countsPerBasin);

        System.out.println(countsPerBasin[numcoords-1] * countsPerBasin[numcoords-2] * countsPerBasin[numcoords-3]);
    }

    public static int basinFill(int x, int y) {
        int[] dx = {1,-1,0,0};
        int[] dy = {0,0,1,-1};

        int sum = 1;
        visited[x][y] = true;

        for (int i=0; i<4; i++) {
            int newx = x+dx[i];
            int newy = y+dy[i];
            if (newx >= 0 && newx < 100 && newy >= 0 && newy < 100 && !visited[newx][newy] && caves[newx][newy] < 9) sum += basinFill(newx,newy);
        }

        return sum;
    }
    
    public static void main(String[] args) throws IOException {
        File f = new File("./AdventOfCode/Y2021/input9.txt");
        BufferedReader reader = new BufferedReader(new FileReader(f));
        String line;
        int linecount = 0;
        while ((line = reader.readLine()) != null) {
            for (int i=0; i<line.length(); i++) caves[linecount][i] = line.charAt(i) - '0';
            linecount++;
        }
        reader.close();

        part1();
        part2();
    }
}

File: 2022/day9.py


def part1():
    visited = {(0,0):True}
    head = (0,0)
    tail = (0,0)
    with open("input9.txt") as f:
        for line in f:
            line = line.split()
            x,y = (1,0) if line[0] == "R" else (0,1) if line[0] == "U" else (-1,0) if line[0] == "L" else (0,-1)
            for i in range(int(line[1])):
                head = (head[0]+x,head[1]+y)
                if tail[0] not in range(head[0]-1,head[0]+2) or tail[1] not in range(head[1]-1,head[1]+2):
                    if tail[0] != head[0] and tail[1] != head[1]:
                        tail = (tail[0] + (1 if tail[0] < head[0] else -1), tail[1] + (1 if tail[1] < head[1] else -1))
                    else:
                        tail = (tail[0]+x,tail[1]+y)
                    visited[tail] = True
    print(len(visited))


def part2():
    visited = {(0,0):True}
    links = [(0,0) for _ in range(10)]
    with open("input9.txt") as f:
        for line in f:
            line = line.split()
            x,y = (1,0) if line[0] == "R" else (0,1) if line[0] == "U" else (-1,0) if line[0] == "L" else (0,-1)
            for i in range(int(line[1])):
                links[0] = (links[0][0]+x,links[0][1]+y)
                for link in range(1,10):
                    if links[link][0] not in range(links[link-1][0]-1,links[link-1][0]+2) or links[link][1] not in range(links[link-1][1]-1,links[link-1][1]+2):
                        if links[link][0] != links[link-1][0] and links[link][1] != links[link-1][1]:
                            links[link] = (links[link][0] + (1 if links[link][0] < links[link-1][0] else -1), links[link][1] + (1 if links[link][1] < links[link-1][1] else -1))
                        elif links[link][0] == links[link-1][0]:
                            links[link] = (links[link][0], links[link][1] + (1 if links[link][1] < links[link-1][1] else -1))
                        else:
                            links[link] = (links[link][0] + (1 if links[link][0] < links[link-1][0] else -1), links[link][1])
                        visited[links[-1]] = True
    print(len(visited))


part1()
part2()
